src/operator/rnn_impl.h

```diff
@@ -176,13 +176,17 @@ void LstmForwardTraining(DType* ws,
       if (dropout > 0.0f) {
         #pragma omp parallel for num_threads(omp_threads)
         for (int j = 0; j < T * N * H * D; j++) {
-          int rand_data = rand_r(&seed_);
+          static thread_local std::random_device device;
```

Comment:
This should probably addressed at the framework level by providing APIs to get rand numbers. We should not expect developers who implement operators to handle thread local variables

Reply:
@eric-haibin-lin In general, I don't see why a thread local variable is an issue - there is a separate generator for every thread and nothing needs to be done additionally to ensure thread safety. Otherwise locking needs to be in place.

I could add a generate random number function to the framework API, but wouldn't change the implementation. What do you think?

Reply:
And in case of 1 function for all there will be complications with setting the seed if needed as well.

Reply:
suggest to refer to `RandGenerator<xpu>` as I mentioned below. It handles thread-safe problem without locking. The implement is somewhat tricky though.